staging/src/k8s.io/code-generator/cmd/validation-gen/validation.go

```diff
@@ -158,10 +190,11 @@ func (g *genValidations) hasValidationsMiss(n *typeNode) bool {
 		allChildren = append(allChildren, n.elem)
 	}
 	for _, c := range allChildren {
-		if !c.fieldValidations.Empty() || !c.keyValidations.Empty() || !c.elemValidations.Empty() {
+		if vc.hasPredicate(c.fieldValidations) || vc.hasPredicate(c.keyValidations) || vc.hasPredicate(c.elemValidations) {
 			return true
 		}
-		if g.hasValidations(c.node) {
+		vc.cache[n] = false // for stack overflow prevention
```

Comment:
@thockin This is a change actual change to the cache logic.

Reply:
> Wouldn't that set cache[T1] = false then recurse into has() on the child, which is typeNode T1, so has() would find that false?

That's right.  And then the parent would continue to evaluate, see `Other string`, return true and update the cache entry to true. The net result is that the recursion is short circuited and the cache entry is set to the correct value by the time `has` returns, but it is temporarily the wrong value..

I'd admit this fix is hacky and the flow get's super twisty.  I could thread in a "visited" set instead if that would help with readability.

Reply:
OK. I think I've addressed this one. I've eliminated hasUpdateVlidations completely from this PR, which simplifies it significantly (IMHO).